### PR TITLE
`tsh request search` support for Pod access requests

### PR DESCRIPTION
**Pull request by tigrato**
This PR adds support for searching, requesting access and approving requests for `kind=pod`.

- Adds `tsh request search --kind pod [--kube-cluster=<kube_cluster>]
  [--kube-namespace=<namespace> | --all-namespaces]`

```
Name                              Namespace Labels                                 Resource ID
--------------------------------- --------- -------------------------------------- --------------------------------------------
nginx-deployment-6595874d85-6j2zm default   app=nginx,pod-template-hash=6595874d85 /tele.teleport.local/pod/minikube/default...
nginx-deployment-6595874d85-7xgmb default   app=nginx,pod-template-hash=6595874d85 /tele.teleport.local/pod/minikube/default...
nginx-deployment-6595874d85-c4kz8 default   app=nginx,pod-template-hash=6595874d85 /tele.teleport.local/pod/minikube/default...
test                              default   run=test                               /tele.teleport.local/pod/minikube/default...

To request access to these resources, run
> tsh request create --resource /tele.teleport.local/pod/minikube/default/nginx-deployment-6595874d85-6j2zm --resource /tele.teleport.local/pod/minikube/default/nginx-deployment-6595874d85-7xgmb --resource /tele.teleport.local/pod/minikube/default/nginx-deployment-6595874d85-c4kz8 --resource /tele.teleport.local/pod/minikube/default/test \
    --reason <request reason>

```

- Adds `tsh request create --resource
  /<cluster>/pod/<kube_cluster>/<namespace>/<podname>`.

Part of #18434
Updates #19573

Built on top of #19613

### CODE REVIEW

File: lib/services/role.go [hunk 1]
```diff
@@ -2101,6 +2101,63 @@ type kubernetesClusterLabelMatcher struct {
...
+// KubeResourcesMatcher matches a role against any Kubernetes Resource specified.
+// It also keeps track of the resources that did not match any of user's roles and
+// that shouldn't be included in the resource ids because the user is not allowed
+// to request them.
+type KubeResourcesMatcher struct {
+	resources     []types.KubernetesResource
+	unmatchedReqs map[string]struct{}
+}
+
+// Match matches a Pod Name name against provided role and condition.
```
Comment (smallinsky) [inline lib/services/role.go:2129]:
```suggestion
// Match matches a Pod name against provided role and condition.
```

File: lib/kube/proxy/forwarder.go [hunk 2]
```diff
@@ -2622,3 +2637,74 @@ func parseDeleteCollectionBody(r io.Reader, decoder runtime.Decoder) (metav1.Del
 	_, _, err = decoder.Decode(data, nil, &into)
 	return into, trace.Wrap(err)
 }
+
+// kubeResourceDeniedAccessMsg creates a Kubernetes API like forbidden response.
+// Logic from:
+// https://github.com/kubernetes/kubernetes/blob/ea0764452222146c47ec826977f49d7001b0ea8c/staging/src/k8s.io/apiserver/pkg/endpoints/handlers/responsewriters/errors.go#L51
+func kubeResourceDeniedAccessMsg(user, method, kubeCluster string, kubeResource *types.KubernetesResource) string {
```
Comment (smallinsky) [inline lib/kube/proxy/forwarder.go:2644]: kubeCluster seems to be unused

File: lib/auth/auth_with_roles.go [hunk 3]
```diff
@@ -1329,6 +1339,208 @@ func (a *ServerWithRoles) ListResources(ctx context.Context, req proto.ListResou
...
+func decideLimit(limit, items int64, respectLimit bool) int64 {
+	if respectLimit {
+		return limit - items
+	}
+	return limit
+}
+
+// getKubernetesClient creates a new Kubernetes client configured with temporary
+// certificates bounded with `search_as_role` roles.
+func (a *ServerWithRoles) getKubernetesClient(ctx context.Context, kubeCluster string) (*kubernetes.Clientset, error) {
```
Comment (smallinsky) [inline lib/auth/auth_with_roles.go:1461]: Can we split the responsibilities of `ServerWithRoles` and move the `getKubernetesClient` kube related functions to other object inside kube package ?

File: lib/services/role.go [hunk 4]
```diff
@@ -2101,6 +2101,63 @@ type kubernetesClusterLabelMatcher struct {
 	clusterLabels map[string]string
 }
 
+// NewKubeResourcesMatcher creates a new KubeResourcesMatcher matcher that
+// matches a role against any Kubernetes Resource specified.
+// It also keeps track of the resources that did not match any of user's roles and
+// that shouldn't be included in the resource ids because the user is not allowed
+// to request them.
+func NewKubeResourcesMatcher(resources []types.KubernetesResource) *KubeResourcesMatcher {
```
Comment (smallinsky) [inline lib/services/role.go:2109]: Could you add test for this matcher ?

**Review by smallinsky — COMMENTED** [review]
Regarding the following UX: 
```
Adds tsh request search --kind pod [--kube-cluster=<kube_cluster>] [--kube-namespace=<namespace> | --all-namespaces]
```

```
Name                              Namespace Labels                                 Resource ID
--------------------------------- --------- -------------------------------------- --------------------------------------------
nginx-deployment-6595874d85-6j2zm default   app=nginx,pod-template-hash=6595874d85 /tele.teleport.local/pod/minikube/default...
nginx-deployment-6595874d85-7xgmb default   app=nginx,pod-template-hash=6595874d85 /tele.teleport.local/pod/minikube/default...
nginx-deployment-6595874d85-c4kz8 default   app=nginx,pod-template-hash=6595874d85 /tele.teleport.local/pod/minikube/default...
test                              default   run=test                               /tele.teleport.local/pod/minikube/default...

To request access to these resources, run
> tsh request create --resource /tele.teleport.local/pod/minikube/default/nginx-deployment-6595874d85-6j2zm --resource /tele.teleport.local/pod/minikube/default/nginx-deployment-6595874d85-7xgmb --resource /tele.teleport.local/pod/minikube/default/nginx-deployment-6595874d85-c4kz8 --resource /tele.teleport.local/pod/minikube/default/test \
    --reason <request reason>
 ```

In case of real use case the number of different pods can be large. We we want list all of them and make the output unreadable. Since the access is requested to all pods in particular namespace but the approval is done per particular pod resource. Does the user that issue access request can identify in easy way to what particle pod resourceID the access request was granted ? 


The code functionality looks good though could you add more test coverage ?

**Review by smallinsky — APPROVED** [review]

**Review by rosstimothy — COMMENTED** [review]
I'm slightly concerned at the precedent this is setting. Up until now I don't believe that there has been a feature that requires Auth to communicate to the Proxy for it to work.

I'm also concerned about latency impacts this will have on anything using ListResources to get pods. If the UI starts to use this it will result in the following:

1. Proxy calls ListResources on Auth
2. Auth lists pods through Proxy
3. Auth returns pods to Proxy

**Comment by tigrato** [comment]
> I'm slightly concerned at the precedent this is setting. Up until now I don't believe that there has been a feature that requires Auth to communicate to the Proxy for it to work.
> 
> I'm also concerned about latency impacts this will have on anything using ListResources to get pods. If the UI starts to use this it will result in the following:
> 
> 1. Proxy calls ListResources on Auth
> 2. Auth lists pods through Proxy
> 3. Auth returns pods to Proxy

@rosstimothy 

Yes, I do agree that this solution isn't ideal but I can't find any valid option without granting access to the cluster itself.

Imagining a scenario where a user does not have access to the cluster but aims to request access to a Kubernetes Pod, we need a process where we can list the pods that Kubernetes Groups and Kubernetes Resource filters allow him to see.
I did consider other solutions such as:
- Kubernetes Service watches the Pods available in the cluster and stores them in Auth Server storage. This is difficult to manage because we do not know the `kubernetes_groups/users` that the user has - i.e. when users use traits provided by a third-party system.  Imagining that we know the Kubernetes principals beforehand, we still need to pick each group's pods, filter them by namespace and apply the Kubernetes RBAC filters on top of it. 
- Users request pods directly to Kube proxy with certificates containing `search_as_roles` roles - this will create a major issue because we will disclose everything available in the cluster.  We can add an extra field to the certificate mentioning that this certificate is only valid for list pods operation - this can be a security concern because `kubectl get pods` will leak pod details other than name and namespace - i.e. env vars that can contain secrets...

I didn't like the two approaches above and opted for this one but I am open to suggestions.

The UI will only use ListPods when the user intends to make an access request.

File: api/proto/teleport/legacy/client/proto/authservice.proto [hunk 5]
```diff
@@ -1681,6 +1683,11 @@ message ListResourcesRequest {
   // UsePreviewAsRoles indicates that the response should include all resources
   // the caller would be able to access with their preview_as_roles
   bool UsePreviewAsRoles = 12 [(gogoproto.jsontag) = "use_preview_as_roles,omitempty"];
+  // KubernetesPodsFilter specifies a Kubernetes Pod filter.
+  types.KubernetesPodsFilter KubernetesPodsFilter = 13 [
```
[the remark this answers is not in the document]
Comment (tigrato) [reply api/proto/teleport/legacy/client/proto/authservice.proto]: Renamed but the type is already defined at `string ResourceType = 1`, so no need to add it again

File: lib/auth/auth_with_roles.go [hunk 6]
```diff
@@ -1329,6 +1338,233 @@ func (a *ServerWithRoles) ListResources(ctx context.Context, req proto.ListResou
...
+	filter := services.MatchResourceFilter{
+		ResourceKind:        req.ResourceType,
+		Labels:              req.Labels,
+		SearchKeywords:      req.SearchKeywords,
+		PredicateExpression: req.PredicateExpression,
+	}
+
+	err := a.iterateKubernetesPods(
+		ctx, req, respectLimit,
+		func(r types.ResourceWithLabels, continueKey string) (int, error) {
```
[the remark this answers is not in the document]
Comment (tigrato) [reply lib/auth/auth_with_roles.go:1369]: Kubernetes Service call does that for us. If the user cannot access the cluster or get pods, the request is eventually denied

File: tool/tsh/tsh.go [hunk 7]
```diff
@@ -885,7 +887,9 @@ func Run(ctx context.Context, args []string, opts ...cliOption) error {
 	reqSearch.Flag("search", searchHelp).StringVar(&cf.SearchKeywords)
 	reqSearch.Flag("query", queryHelp).StringVar(&cf.PredicateExpression)
 	reqSearch.Flag("labels", labelHelp).StringVar(&cf.UserHost)
-
+	reqSearch.Flag("kube-cluster", "Kubernetes Cluster to search for Pods").StringVar(&cf.KubernetesCluster)
+	reqSearch.Flag("kube-namespace", "Kubernetes Namespace to search for Pods").Default(corev1.NamespaceDefault).StringVar(&cf.kubeNamespace)
+	reqSearch.Flag("all-namespaces", "Search Pods in every namespace").BoolVar(&cf.kubeAllNamespaces)
```
Comment (jakule) [inline tool/tsh/tsh.go:892]: For me, naming is inconsistent. On the one hand, you have `--kube-namespace` that suggest K8S namespace, but then `--all-namespaces` sounds confusing as the name doesn't says anything about which namespaces you are referring to.

**Comment by smallinsky** [comment]
Regrading:

> > I'm slightly concerned at the precedent this is setting. Up until now I don't believe that there has been a feature that requires Auth to communicate to the Proxy for it to work.
> > I'm also concerned about latency impacts this will have on anything using ListResources to get pods. If the UI starts to use this it will result in the following:
> > 
> > 1. Proxy calls ListResources on Auth
> > 2. Auth lists pods through Proxy
> > 3. Auth returns pods to Proxy
> 
> @rosstimothy
> 
> Yes, I do agree that this solution isn't ideal but I can't find any valid option without granting access to the cluster itself.
> 
> Imagining a scenario where a user does not have access to the cluster but aims to request access to a Kubernetes Pod, we need a process where we can list the pods that Kubernetes Groups and Kubernetes Resource filters allow him to see. I did consider other solutions such as:
> 
> * Kubernetes Service watches the Pods available in the cluster and stores them in Auth Server storage. This is difficult to manage because we do not know the `kubernetes_groups/users` that the user has - i.e. when users use traits provided by a third-party system.  Imagining that we know the Kubernetes principals beforehand, we still need to pick each group's pods, filter them by namespace and apply the Kubernetes RBAC filters on top of it.
> * Users request pods directly to Kube proxy with certificates containing `search_as_roles` roles - this will create a major issue because we will disclose everything available in the cluster.  We can add an extra field to the certificate mentioning that this certificate is only valid for list pods operation - this can be a security concern because `kubectl get pods` will leak pod details other than name and namespace - i.e. env vars that can contain secrets...
> 
> I didn't like the two approaches above and opted for this one but I am open to suggestions.
> 
> The UI will only use ListPods when the user intends to make an access request.

@tigrato

Does the Auth cannot expose a Method that will generate a cert with for `search_as_roles` and for the teleport proxy so teleport proxy will use this cert in kube client to list resources ?

**Comment by fspmarshall** [comment]
I'm also concerned about the current model of auth indirectly dialing the Kubernetes API.  In addition to it just being fairly indirect (`clt -> proxy -> auth -> proxy -> agent -> k8s`), it is an unprecedented shift in how teleport's networking works.  Teleport does no currently require auth servers to be able to dial _any_ other member of the teleport cluster.  Furthermore, hiding a complex remote call like this inside of `ListResources` is problematic IMO, since that method is otherwise a simple cache lookup mechanism.


@tigrato Based on your discussion of alternative impls [here](https://github.com/gravitational/teleport/pull/19791#issuecomment-1382143171), this alternative seems interesting:

> Users request pods directly to Kube proxy with certificates containing search_as_roles roles - this will create a major issue because we will disclose everything available in the cluster. We can add an extra field to the certificate mentioning that this certificate is only valid for list pods operation - this can be a security concern because kubectl get pods will leak pod details other than name and namespace - i.e. env vars that can contain secrets...

It sounds to me like the main downside here is that we are forced to perform the "search as roles" operation in a manner that masquerades as a normal k8s API call.  The agent itself is perfectly capable of serving the request without auth as an intermediary, but because we only communicate with it _as if_ it were a k8s apiserver, we have to do hacky stuff with certs and worry about filtering out secrets.

Speaking hypothetically, if we had the ability to call methods and/or send messages to agents _outside_ of the normal k8s protocol, then this problem would be trivial, right?  The agent could just expose a custom `SearchPodsAsRoles` method/message, and tsh could dial the agent and get the required info just as easily as listing pods normally, but with none of the concerns of juggling certs and/or filtering sensitive info.

Assuming the above is correct, I think this is actually just the latest example of a problem we've seen before... passing sideband information through the revrsetunnel system is difficult.  We need a way to add custom messages/methods outside of the scope of an agent's particular protocol.

Say, for example, if we extended the strategy from the proxy GRPC transport proposal (https://github.com/gravitational/teleport/pull/19439) that @rosstimothy did recently.  If we moved to GRPC-based `proxy <-> k8s-agent` tunnels, then we could introduce new message variants that let us add new agent interactions any time we wanted.  Ex:

```
message KubernetesProxyRequest {
  Frame k8s_frame = 1; // standard k8s protocol, as is currently proxied
  SearchPodsAsRoles search_pods = 2; // perform a SearchAsRoles directly against the agent
  // ...
}
```

With something like the above, "search pods as role" becomes a standard `clt -> proxy -> agent` operation, no different than a standard list pods, except that we're using a custom message instead of proxying the native k8s protocol.

I'm not sure what the timeline is on landing pod access requests, but I strongly believe that we're better off in the long run if we solve the general problem of making agent interaction extensible, rather than trying to inject auth as a middle man.

**Comment by tigrato** [comment]
@fspmarshall @rosstimothy 

I subscribe to your concerns and dropping `auth` as a middle-man is something that makes sense. 

From the request perspective, adding a different endpoint - let's call `list_pods_as_roles` for the time being - solves the leaking env secrets and allows the agent to return a `KubernetesPodV1` list. In that case, I think we can append the `search_as_roles` directly into Kubernetes Service Auth Checker. This solution also requires bypassing Auth Server from returning a certificate that embeds the Kubernetes cluster name into `identity.KubernetesCluster` - for cases where the user's roles don't allow access to the cluster. 

Probably, the `list_pods_as_roles` endpoint can receive the target cluster name from a query arg or even the request body and Kube Proxy forwards the request to Kube Service. On the other hand, the Kube Service will validate the existence of the target cluster and check if the user's roles + `search_as_roles` allow access to the target cluster and filter the pods based on `kubernetes_resources` rules, returning the pods as `KubernetesPodV1` - only shows the name, namespace, and labels. 

This solution will allow you to directly use the user's credentials - without the user `ca.Identity` embedding `kubernetes_groups` and `kubernetes_cluster` - for `lust_pods_as_roles` endpoint, but it forces local validation to take place. It also prevents the Auth Server from leaking certificates to the user since everything is generated and validated by the Kube Proxy/Service, and the user never receives them.

However, the extension of the solution described in the RFD 100 would allow more configuration options in a safer way than what is allowed by the version I described above.

Besides the written above, I think that extending RFD 100 will solve another issue we have with Kube Proxy. Currently, the Kube proxy re-generates a client certificate on behalf of the user and caches it locally for `min(cert_expiration,3h)` in order to pass the user identity to the Kubernetes Service since the proxy does TLS termination. Besides the latency that certificate generation introduces, it has created several issues (#18070, #19884)  in the past few months because the credentials cache was using incorrect keys in the proxy and once the user calls with a different certificate, it reuses the wrong certificate until the cache expires - the root cause was missing parameters in the cache's [key](https://github.com/gravitational/teleport/blob/v8.3.20/lib/kube/proxy/forwarder.go#L321-L325). Allowing the Proxy to connect to Kube Service using an extension of RFD 101 would allow it to pass the user's identity directly without requiring generating a new certificate and caching it locally and, eventually, fix all caching issues and latency introduced when generating new certificates.


What do you think?

**Comment by tigrato** [comment]
In order to include some of this PR's functionality in Teleport 12, I created PR #20272 which adds support for `tsh request create`. Support for `tsh request search --kind=pod` will be introduced later once we agree

**Comment by tigrato** [comment]
Closing since the same feature was implemented by #21103 and #21107

**Pull request closed without merging**